src/net/larse/lcms/algorithms/VCT.java

```diff
@@ -0,0 +1,1567 @@
+              double[] distMagnB4) {
+        this.years = Ints.asList(Arrays.copyOfRange(years, 0, nYears));
+        if (lcType != PART_FOREST) {
+          this.distFlag = new ArrayList<>(Collections.nCopies(nYears, lcType));
+          this.distMagn = new ArrayList<>(Collections.nCopies(nYears, 0.0));
+          this.distMagnVi = new ArrayList<>(Collections.nCopies(nYears, 0.0));
+          this.distMagnBr = new ArrayList<>(Collections.nCopies(nYears, 0.0));
+          this.distMagnB4 = new ArrayList<>(Collections.nCopies(nYears, 0.0));
+        } else {
+          this.distFlag = Ints.asList(Arrays.copyOfRange(distFlag, 0, nYears));
+          this.distMagn = Doubles.asList(Arrays.copyOfRange(distMagn, 0, nYears));
+          this.distMagnVi = Doubles.asList(Arrays.copyOfRange(distMagnVi, 0, nYears));
+          this.distMagnBr = Doubles.asList(Arrays.copyOfRange(distMagnBr, 0, nYears));
+          this.distMagnB4 = Doubles.asList(Arrays.copyOfRange(distMagnB4, 0, nYears));
+        }
+      }
+    }
+  }
+
+  /**
+   * Class for retaining information about segments across a time series
+   */
+  public static class TSSegment {
+
+    public int segType;         // type for this segment
+    public int segLength;       // length for this segment
+    public int startYear;       // start year for this segment
+    public int endYear;         // end year for this segment
+
+    public TSSegment(int segType, int startYear, int segLength) {
+      this.segType = segType;
+      this.startYear = startYear;
+      this.endYear = startYear + segLength - 1;
+      this.segLength = segLength;
+    }
+  }
+
+  /**
+   * Given a time series and two endpoints, linearly interpolate all values
+   *
+   * @param ts - Array of values to interpolate
+   * @param left - Left endpoint to use
+   * @param right - Right endpoint to use
+   */
+  public static final void interpolateValues(double[] ts, int left, int right) {
+    double denom = (double) (right - left);
+    double slope = (ts[right] - ts[left]) / denom;
+    for (int j = left + 1; j < right; j++) {
+      ts[j] = ts[left] + slope * (j - left);
+    }
+  }
+
+  /**
+   * Calculate a measure of a time series' inter-annual variability
+   *
+   * @param ts - Array of time series values
+   * @param left - Left endpoint to use
+   * @param right - Right endpoint to use
+   * @return - Roughness value
+   */
+  public static final double fiRoughness(double[] ts, int left, int right) {
+    int numVals = right - left + 1;
+
+    if (numVals <= 3) {
+      return -1.0;
+    }
+
+    // Find all differences between consecutive data pairs
+    double[] tmpData = new double[numVals - 1];
+    for (int i = left; i < right - 1; i++) {
+      tmpData[i - left] = ts[i + 1] - ts[i];
+    }
+
+    // Sort these differences and find the index 1/10 in from the left
+    // constrained betweeen indexes 1 and 3 inclusive; 
+    // return the absolute value of this pair's difference
+    Arrays.sort(tmpData);
```

Comment:
Make sure you use Arrays.sort(float[] a, int fromIndex, int toIndex) when you move to a fixed length, reusable array.